### ISSUE TITLE
Need a `KN_NAMESPACE` env variable to avoid repeating same namespace

### ISSUE THREAD

**Issue by maximilien** [knative/client#184]
While restoring the e2e tests that use `bash` I realize that a pain point that users of `kn` will experience is having to repeat the same namespace when issuing many commands in a row---If that namespace is not the default namespace. 

See the PR #183 or sample snippet here to illustrate:

```bash
# Listing 1
# Will create and delete this namespace and use it for smoke tests
export KN_E2E_SMOKE_TESTS_NAMESPACE=kne2esmoketests

kubectl create ns $KN_E2E_SMOKE_TESTS_NAMESPACE

./kn service create hello --image gcr.io/knative-samples/helloworld-go -e TARGET=Knative -n $KN_E2E_SMOKE_TESTS_NAMESPACE
sleep 5
./kn service get -n $KN_E2E_SMOKE_TESTS_NAMESPACE -n $KN_E2E_SMOKE_TESTS_NAMESPACE
./kn service update hello --env TARGET=kn -n $KN_E2E_SMOKE_TESTS_NAMESPACE
sleep 3
./kn revision get -n $KN_E2E_SMOKE_TESTS_NAMESPACE
./kn service get -n $KN_E2E_SMOKE_TESTS_NAMESPACE
...
```

A better option would be to have `kn` use `KN_NAMESPACE` env value as the namespace for commands if that's defined. 

So above becomes cleanly:

```bash
# Listing 2
# Will create and delete this namespace and use it for smoke tests
export KN_NAMESPACE=kne2esmoketests

./kn service create hello --image gcr.io/knative-samples/helloworld-go -e TARGET=Knative
sleep 5
./kn service get
./kn service update hello --env TARGET=kn
sleep 3
./kn revision get
./kn service get
...
```

**Comment by duglin** [comment]
I'm not against a KN_NAMESPACE env var, but I'd like it if we also picked up the namespace from the kube config. So, order of processing would be:
1 - `-n` flag
2 - `KN_NAMESPACE`
3 - Namespace from kube config

Then if people are switching between `kn` and `kubectl`, setting just the kube one will work for both.

**Comment by maximilien** [comment]
Yup. That's reasonable.

**Comment by rhuss** [comment]
@maximilien that is already fixed with #36 and using an intial `kubectl config use-context` I don't see the need for a special KN_NAMESPACE. Just use the current context's namespace by default.

**Comment by maximilien** [comment]
Hmm, not clear. Maybe I need to understand context more. 

However, how would you convert the snippet above (listing 1) to look like the second one. That is avoid repeating the `-n $KN_E2E_SMOKE_TESTS_NAMESPACE` in each command. Is it create a context, create namespace. Make that default? 

Remember `-n` is an option in all commands. I suspect most users will do what I listed above (listing 1). Thoughts?

**Comment by duglin** [comment]
I think he means:
```
# permanently save the namespace for all subsequent kubectl commands in that context.
kubectl config set-context --current --namespace=ggckad-s2
```
from: https://kubernetes.io/docs/reference/kubectl/cheatsheet/

will set it once for all future kubectl/kn commands.
One of the reasons I was (sort of) ok with KN_NAMESPACE is because I think trying to remember (or having to alias) that kubectl config cmd is asking too much :-)  Why there isn't a KUBE_NAMESPACE env var.... I dunno.

**Comment by maximilien** [comment]
How is that a great user experience? At least creating a namespace with kubectl is easy: `kubectl create ns blah`

With the config/context scheme, the user has to be using context, understand them, and set it to use the namespace she has in mind. Then remember to change context or namespace of context when she wants to use a different namespace for a series of commands.

I don't see this as a great UX. Especially since we sprinkle `-n` option to all commands. If we also had a global context option then I could see that we are suggesting users to consider context. But AFAIK we don't.

Anyhow, I've said my .02.

**Comment by rhuss** [comment]
You can see the 'current context' like a current working directory. Every UNIX command happens in that context (ls, touch, cat), except when you use an absolute path (in our analogy when you use `-n`). 

Except that changing the current context is a bit tedious (with `kubectl config use-context <myns>`) its exactly the same concept. There are tools like [kubens](https://github.com/ahmetb/kubectx) which help here. OpenShift's `project` (which is the extension to a namespace) can be easily switched which `oc project <news>` (and that is done all the time).

Having a new mechanism to switch that context with a new env variable is not a good UX as is very implicit. 

I agree that 

```
export KN_NAMESPACE=e2e-test
```

is better than

```
kubectl config set-context $(kubectl config current-context) --namespace=e2e-test
```

as described [here](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#setting-the-namespace-preference), but its also very implicit.

Maybe it's now time to introduce a `~/.kn` config directory/file which may hold either a 'default_namespace' file which can be set with `kn set default namespace=....` (this could e.g. also use for setting the default wait behaviour like with `kn set default wait=false`).

However, its also important to note, that when we rely on `~/.kube/config` as in #172 we introduce a subtle dependency on Kubernetes client tooling (which I think kn might want to try to avoid). We have that already for authentication, not sure how much this is "acceptable".

**Comment by rhuss** [comment]
> With the config/context scheme, the user has to be using context, understand them, and set it to use the namespace she has in mind. Then remember to change context or namespace of context when she wants to use a different namespace for a series of commands.

Isn't this the same when using an environment variable, so that you have to remember to change the environment variable when you want to use a different namespace ?

What I was suggesting is not to introduce a new context each time but to update the current context with the target namespace (like kubens does)

**Comment by duglin** [comment]
How about we add this:
```
kn set-namespace foo
# or
kn config namespace=foo
# which leave it open for other types of config options
```
then these can do the `kubectl config set-context $(kubectl config current-context) ...` under the covers in the case of setting the NS. Or when we're not setting the NS it sets some `$HOME/.kn/config` thingy as needed.

The point here for me is that it would be really nice if, as a user, I didn't have to set two different NS config options because I expect people to jump between `kn` and `kubectl` a lot.

**Comment by maximilien** [comment]
Any other opinions on this issue? I am happy to implement the last suggestion.

@rhuss @sixolet @cppforlife chime in if you feel strongly about alternative approaches.

**Comment by navidshaikh** [comment]
Once https://github.com/knative/client/pull/172 is in, we'll just need to create and set the desired namespace once and we should be good. No need for another var for namespace reference.

**Comment by rhuss** [comment]
I think the issue is, how we can easily set the current namespace directly with `kn` and without extra help from `kubectl` or `kubens`.

So having a 

```
kn set-namespace food
```

as an alternative to 

```
kubectl config set-context $(kubectl config current-context) --namespace=foo
```

sound a good idea as long as both `kn` and `kubectl` work on the same config so that they always both use the same default namespace. Nothing would be more confusing than having kubectl another default namespace than kn.

**Comment by navidshaikh** [comment]
> kn set-namespace food

This seems reasonable as long as the desired namespace is recorded in kubeconfig (and not in some env var).

**Comment by maximilien** [comment]
OK waiting on @sixolet and/or @cppforlife for chime in. I am cool with `kn set-namespace food`

Any other opinions? 

----
FYI to all... Nǐ hǎo 网络你好 from Shanghai for KubeCon/CNCon where many internet things don't work :( but GitHub does :) Otherwise all good but expect odd delays in responses.

**Comment by maximilien** [comment]
OK still have not heard from @sixolet and @cppforlife on this. Any opinions?

Otherwise, I can cook up a PR for this. Though we should think about how to retrieve the namespace and removing it. So perhaps something like:

```bash
kn namespace set lol
default namespace set to "lol"

kn namespace get
default namespace is "lol"

kn namespace unset
default namespace is unset

kn namespace get
default namespace is not set
```

**Comment by rhuss** [comment]
Agree to use a dedicated `config` command, instead of "wasting" a group for the namespace only (we should also strive to not pollute this top-level commandgroup namespace too much).

We might also consider the `kubectl` way (I know, not all is good with kubectl UX, but I think this is not so bad, too):

```
# Set the default namespace
kn config set current-namespace foo

# Clear namespace
kn config unset current-namespace

# Get a single config variable
kn config get current-namespace

# View whole config, not only the namespace
kn config view
```

(whether its `namespace`, `default-namespace`, `current-namespace` is probably not so important, but would like to have it a bit more specific, especially in the case when we should add more options)

The good thing about this approach is, that its a single UI for configuration which might come from different sources. Like the namespace is stored in `~/.kube/config`, but other config options could come from `~/.kn/config`

**Comment by duglin** [comment]
I tend to prefer a "config" command as well

**Comment by rhuss** [comment]
Should we create a new issue for allowing advanced configuration management ? (like managing the namespace, context but also our own config like listing, creating, deleting mappings via scripting ?)

Especially managing sink prefixes via a `kn config add sink-prefix svc --group ... --kind ... --version` or so might be helpful in longer scripts in e.g. a CI which always starts with an empty config.

// @navidshaikh @maximilien @duglin
